src/nvim/lua/treesitter.c

```diff
@@ -1411,26 +1417,44 @@ static int query_next_capture(lua_State *L)
     push_node(L, capture.node, lua_upvalueindex(2));  // [index, node]
 
     // Now check if we need to run the predicates
+    // a predicate is executed only once within the same match.
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
 
-    if (n_pred > 0 && (ud->max_match_id < (int)match.id)) {
+    if (n_pred == 0) {
+      // no predicates. Return without the match table as a small perf optimization,
+      // because in the iter_captures() method, match table won't be needed.
+      // [index, node]
+      return 2;
+    }
+
+    if (ud->max_match_id < (int)match.id) {
       ud->max_match_id = (int)match.id;
 
-      lua_pushvalue(L, lua_upvalueindex(4));  // [index, node, match]
+      // Create a new cleared match table
+      lua_createtable(L, (int)ts_query_capture_count(query), 2);  // [index, node, match]
+      // match: TSMatch = { [integer]: TSNode, pattern: integer, active: boolean }
       set_match(L, &match, lua_upvalueindex(2));
+      // match.pattern
       lua_pushinteger(L, match.pattern_index + 1);
       lua_setfield(L, -2, "pattern");
 
+      // Set match.active = false for the next match.
       if (match.capture_count > 1) {
         ud->predicated_match = (int)match.id;
         lua_pushboolean(L, false);
         lua_setfield(L, -2, "active");
       }
-      return 3;
+
+      // Update the current_match in upvalue with the new match table
+      lua_replace(L, lua_upvalueindex(4));  // [index, node]
     }
-    return 2;
+
+    lua_pushvalue(L, lua_upvalueindex(4));  // [index, node, match]
+    return 3;
   }
+
+  // end of result for the query.
```

Comment:
I think this whole thing can be simplified to always include the `match` table and return 3 arguments unconditionally. The `iter_captures` method in `query.lua` will need to be updated appropriately to avoid re-evaluating predicates and directives multiples times per match, though (see the test case I provided to check for this).

The first time a match is visited we set `match.active` to false. I think we can use this in the `iter_captures` implementation to determine when we need to evaluate `match_preds` and `apply_directives`.

Reply:
Thanks for sharing the test case. I will try to include them -- probably moving query related tests out of `parser_spec` (which is now too long) to a new `queries_spec`.

Yes, in this context predicates includes directives (which is a Nvim-specific term). A clarification would be needed in the docs. While returning only when there is a predicate/directive would be fine, if we later decide to make `iter_capture()` return the fourth return value `match` (for #27132) then this might need to be reconsidered; e.g. via opt-in.

Reply:
For #27312 at least, we only need `match` when there are directives, so this would still be ok.